### ISSUE TITLE
Neural Net Math Learning 

### ISSUE THREAD

**Issue by Rohan2821999** [Rohan2821999/MathCog_Modelling#3]
@cbattista Sir,  I was trying to examine the relationship between the cost function (error) and no. of runs. The  graph is given below -- learning rate of the gradient descent used is 0.1 : 

![image](https://cloud.githubusercontent.com/assets/13100688/18332125/f67d474c-7518-11e6-9267-6e6fce0ffc1b.png)

I think this cost function is kind of matching human learning because initially the artificial neuron is learning faster (when its badly wrong)  and learns slower as error decreases. However, at cost 0.2 and epochs 2000 it kind of becomes asymptotic and no longer learning. 

Do you think this would be the ideal cost function for our neural net?

**Comment by Rohan2821999** [comment]
Sorry for the typing error, I mean at cost 0.2 and epochs 7000 it becomes asymptotic

**Comment by Rohan2821999** [comment]
I ran the neural net for just addition and it works perfect (100% Accuracy) after 1000 runs.. However, I noticed that the output values are still not 1s or 0s instead its <= 0.39 for 0s and around 0.41 for all 1s.. This means that the values are correct just haven't been scaled well. So the neural net is working fine. 

I re-tested the neural net on all data (addition and subtraction) -- accuracy - 70% however, again, the values are not 0s or 1s rather other values which definitely do follow a desirable and more or less correct trend..

**Comment by cbattista** [comment]
OK, so at least we can teach it addition!  But yeah the cost/output
function needs to be updated.  Spoke with John, and he recommends a
cross-entropy function for this, so try that - basically it's doing a
similar thing as logistic regression.

-C

**Comment by Rohan2821999** [comment]
So, I didn't use the cross entropy function because the learning graph of that is super weird. Also, the values after that are also not discretized as 0 or 1 (same as before):

Graph of cross entropy function vs number of runs:
![image](https://cloud.githubusercontent.com/assets/13100688/18363266/3a472786-75be-11e6-9b99-65772c519916.png)

I am not clear how the learning is taking place here!

**Comment by Rohan2821999** [comment]
Update on NN stuff: 
1. I was previously wrong about the subtraction learning in NN. The NN is learning both addition and subtraction with complete accuracy for all numbers. The only place where it makes a mistake is when the subtraction computation is 0 (eg 5-5) since that cannot be correctly represented by the softmax model (since the model values are probabilistic and should add up to 1). So the model is not wrong and working with complete accuracy. 
2. The model learns pretty quickly (24 runs) to get a 0.979 accuracy and then becomes asymptotic. Below is the graph for the cost function for this:

![type1datacost](https://cloud.githubusercontent.com/assets/13100688/18369427/840351e4-75da-11e6-86f0-1c8e7bc09975.JPG)
1. Regarding unsupervised learning none of the models (addition or subtraction + addition) work for new test data. Has an accuracy of 0.4 with new data. This could probably be because the training data samples are low. I am unsure about how to get this part working. 
2.  I tried using the current model for the  2nd type of data but the biases and hypothesis values become NaN after the third run. Unsure about why this is happening.

**Comment by Rohan2821999** [comment]
Implemented Neural Net on only addition vals for type3 data. Doesn't work at all. No learning takes place even after 500 runs and accuracy then is absolutely 0.

**Comment by cbattista** [comment]
Awesome.
